### DISCUSSION TITLE
QueuePool limit hit even though PostgreSQL doesn't show any connections held

### DISCUSSION THREAD

**Discussion by lucas03** [sqlalchemy/sqlalchemy#10955] [Usage Questions]
Hey,
I wanted to create an issue, but as mentioned there, discussion might be a better place to start. Unfortunately I don't have any reproducable code for this.
For past half or year or so, we have been alerted by increase in these errors:
```
TimeoutError('QueuePool limit of size 4 overflow 20 reached, connection timed out, timeout 30.00')
```
In logs, it looks like this:
<img width="660" alt="image" src="https://github.com/sqlalchemy/sqlalchemy/assets/2773622/9a6591f9-1d40-4d54-9f25-07da9a518e7f">
950 unique requests failed, out of 7140 requests within the same time range (uniformly spread).

TimeoutError is being raised until somebody comes and restarts affected pod. It always happens on one pod only, even though we have like 100 gunicorn pods running simultaneously (maybe like 6 different deployments). It randomly occurs on all of them.
I am aware of [this page](https://docs.sqlalchemy.org/en/20/errors.html#connections-and-transactions), but this seemed to me like a bug. We hoped it might be resolved with future SQLAlchemy version, but it's been happening since 17.7.2023 and still strong.

We tried to investigate, what connections are being held at the time, but pg_stat_activity (managed PostgreSQL 13 from google) showed none (there were just 4 idle connections).

We use connexion==2.14.2 (flask==2.2.5) with gevent==23.9.1 and sqlalchemy==2.0.23 (was happening on previous version of SQLAlchemy as well).

We do monkey patch gevent like this on app start:

```py
    if gevent:
        from gevent import monkey  # isort:skip

        monkey.patch_all()

        from grpc.experimental.gevent import init_gevent

        init_gevent()

        from psycogreen.gevent import patch_psycopg  # isort:skip

        patch_psycopg()
```
we create session like this:
```py
def _init_db(
    url: URL | str,
    name: str,
    statement_timeout: int | None = None,
) -> ScopedSession:
    """Create and initialize an engine and a session to booking database.

    :param url: url style connection string
    :param name: connection name in tracers
    :param statement_timeout: optional number of seconds to set db statement_timeout
    :return: booking session.
    """
    psql_app_name = f"{settings.environment} | {socket.gethostname()} | {settings.app_name}"
    psql_app_name = psql_app_name[:63]  # psql limit is 63 characters

    connect_args = {"application_name": psql_app_name, "connect_timeout": 10}

    pg_options: list = [f"idle_in_transaction_session_timeout={settings.pg_idle_in_transaction_session_timeout}"]

    if statement_timeout is not None:
        pg_options.append(f"statement_timeout={statement_timeout * 1000}")

    if pg_options:
        connect_args["options"] = f'-c {" ".join(c for c in pg_options)}'

    engine = create_engine(
        url,
        connect_args=connect_args,
        pool_recycle=1800,  # 30 minutes, https://cloud.google.com/sql/docs/postgres/manage-connections#duration
        pool_size=settings.pg_pool_size,  # maximum connections
        max_overflow=settings.pg_max_overflow,
        json_serializer=json.dumps,
        pool_pre_ping=True,
    )

    for event_name in ("first_connect", "checkout", "checkin"):
        event.listen(engine, event_name, partial(_log_pool_event, engine, event_name))

    event.listen(engine, "connect", partial(_on_db_connect, name))

    if settings.app_name in settings.logging.enabled_connection_events:
        connection_logger = ConnectionLogger()
        event.listen(engine, "checkout", connection_logger.log_checkout)
        event.listen(engine, "checkin", connection_logger.log_checkin)

    session_ = sessionmaker(bind=engine, future=True)
    session = scoped_session(session_)

    return session
```
Any tips how to fix this issue? Since I didn't see any connections held in DB for affected pods, I think it's either SQLAlchemy bug or our misuse in app.

**Comment by zzzeek** [comment]
how are you measuring the actual connections to postgresql?  looking at the server side for PG is fine, that will show you what the server sees.  but what does the client see?  are there 24 TCP connections that were set up and are timing out, and that would be the checked out connections?      do you have 24 requests in your app that are already waiting, and if so what are they waiting on ?   

unfortunately mixing up gunicorn with gevent plus monkeypatching creates a whole lot of possibilities for weird things to happen so I think you have to look in your logs and see what happens before you see this error, do you see requests failing, were connections cut off, etc, there's a lot happening here.   

you are correct in not opening a bug because how would we go about doing anything here, unfortunately you will have to figure out what triggers this and make more detailed observations of the state of the application when this happens.   short of that I'd seek to move away from gevent monkey patching which is generally a recipe for big problems.

**Comment by lucas03** [comment]
Thank you for a swift reply, appreciate your work and hope you are doing well.

> how are you measuring the actual connections to postgresql?

`SELECT * FROM pg_stat_activity WHERE application_name = '{affected_pod_name}'`.

>  do you have 24 requests in your app that are already waiting, and if so what are they waiting on ?

We enabled logs for checkin and checkout and optimized a few cases, but it didn't help with random QueuePool errors. As they happen in any place in codebase, I don't think it's caused by long lasting calls. (We don't have that long API calls and also optimize slow SQL queries). I have been checking number of requests or specific endpoints, but there wasn't peak or any obvious pattern. Multiple different paths as well.

> how would we go about doing anything here

Yeah, I understand. But as we are out of ideas, I thought you might point us to some setting or known issue. :(

**Reply by zzzeek** [reply]
the issues with gevent / eventlet in the past typically stem from when there's a legit problem with a connection, or the greenlet it's running in, which can be the connection itself is severed, the client receiving data from the greenlet closes their browser and you get a CancelledError, things like that; and then from that point, the application / SQLAlchemy /etc does not respond correctly and the state of the pool gets messed up as a result.   This is why it's important to figure out what kind of anomalous thing precedes the event.

the other possibility is some kind of race condition, which would be a little more ominous to figure out as those aren't preceded by anything interesting.

**Reply by lucas03** [reply]
okay. Since gunicorn recommends gevent for async, out of curiosity, what do you recommend as tech stack for sqlalchemy here? Asyncio?

**Reply by zzzeek** [reply]
I've just never understood gunicorn's concurrency model.   So, they want you to gevent patch your app, OK that's fine, and then you tell gunicorn that it should spawn gevent workers?  is that how it works?

I mean yeah asyncio is less "mysterious" and more explicit in the API, but the execution model as far as "what runs when, how are things scheduled" will look extremely similar.    Mostly if you rewrote your app in asyncio, then whatever pattern is causing the problems now would just have gone away by virtue of reworking the code.

I dont think it's time to rewrite your app unless it's fairly trivial, I'd still be looking to determine a. what is the precipitating event and b. when it happens, getting a more complete picture of what's going on inside the worker

**Reply by zzzeek** [reply]
also , it's too early to talk about this, but there's a [real gil removal](https://peps.python.org/pep-0703/) effort underway which I think at least in the short to medium term is going to turn all of this on its head; threading will offer true CPU parallelism and asyncio / gevent will not.

There is also still my original assertion that performance with non-blocking IO vs. blocking IO is not that different in Python, personally I would stick to using Apache / mod_wsgi / threads with processes.   You'd have a very stable and predictable application.

**Reply by lucas03** [reply]
hey, to report back:

- We tried upgrading psycopg2 to psycopg. Error still occured.
- Later we have switched from gevent gunicorn worker to gthread. It's been two weeks and we haven't noticed these errors anymore. Downside is gthread has almost twice memory consumption and it's more volatile (memory jumps up violently and it's less predictable).

**Reply by zzzeek** [reply]
is this something you can reproduce with gevent in a controlled / test scenario?   like hitting the application with a web testing tool?

**Reply by lucas03** [reply]
not really, we had 8 services with ~30M requests per day, and this error happened just once a week or so.
